core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -813,18 +813,19 @@ private[zio] final class FiberContext[E, A](
   def id: Fiber.Id = fiberId
 
   def inheritRefs: UIO[Unit] = UIO.effectSuspendTotal {
```

Comment:
I'm not sure this is buying us anything. `locals` is our own snapshot of the `fiberRefLocals` and we are free to work with `locals` as much as we want without any synchronization. We just need to synchronize when we create our snapshot to make sure we have a consistent view.

So in the current implementation we are always synchronizing exactly once when we do `Sync` on line 820. In contrast, in the proposed implementation we synchronize either once or twice. We always synchronize once when we do `fiberRefLocals.isEmpty` since as you say this synchronizes internally and then whenever the result is not empty we synchronize a second time to actually create the snapshot of the values.

So I tend to think this is not an improvement over the current implementation.